tsl/src/remote/connection.c

```diff
@@ -170,6 +171,147 @@ typedef struct TSConnection
  */
 static ListNode connections = { &connections, &connections };
 
+static bool
+fill_simple_error(TSConnectionError *err, int errcode, const char *errmsg, const TSConnection *conn)
+{
+	if (NULL == err)
+		return false;
+
+	MemSet(err, 0, sizeof(*err));
+
+	err->errcode = errcode;
+	err->msg = errmsg;
+	err->host = pstrdup(PQhost(conn->pg_conn));
+	err->nodename = pstrdup(NameStr(conn->node_name));
+
+	return false;
+}
+
+static bool
+fill_connection_error(TSConnectionError *err, int errcode, const char *errmsg,
+					  const TSConnection *conn)
+{
+	if (NULL == err)
+		return false;
+
+	fill_simple_error(err, errcode, errmsg, conn);
+	err->connmsg = pstrdup(PQerrorMessage(conn->pg_conn));
+
+	return false;
+}
+
+static char *
+get_error_field_copy(const PGresult *res, int fieldcode)
+{
+	const char *msg = PQresultErrorField(res, fieldcode);
+
+	if (NULL == msg)
+		return NULL;
+	return pchomp(msg);
+}
+
+/*
+ * Convert libpq error severity to local error level.
+ */
+static int
+severity_to_elevel(const char *severity)
+{
+	/* According to https://www.postgresql.org/docs/current/libpq-exec.html,
+	 * libpq only returns the severity levels listed below. */
+	static const struct
+	{
+		const char *severity;
+		int elevel;
+	} severity_levels[] = { {
+								.severity = "ERROR",
+								.elevel = ERROR,
+							},
+							{
+								.severity = "FATAL",
+								.elevel = FATAL,
+							},
+							{
+								.severity = "PANIC",
+								.elevel = PANIC,
+							},
+							{
+								.severity = "WARNING",
+								.elevel = WARNING,
+							},
+							{
+								.severity = "NOTICE",
+								.elevel = NOTICE,
+							},
+							{
+								.severity = "DEBUG",
+								.elevel = DEBUG1,
+							},
+							{
+								.severity = "INFO",
+								.elevel = INFO,
+							},
+							{
+								.severity = "LOG",
+								.elevel = LOG,
+							},
+							/* End marker */
+							{
+								.severity = NULL,
+								.elevel = 0,
+							} };
+	int i;
+
+	if (NULL == severity)
+		return 0;
+
+	i = 0;
+
+	while (NULL != severity_levels[i].severity)
+	{
+		if (strcmp(severity_levels[i].severity, severity) == 0)
+			return severity_levels[i].elevel;
+		i++;
+	}
+
+	pg_unreachable();
```

Comment:
Macro `pg_unreachable` is intended for cases where the full situation cannot be caught by the compiler, for example, after a statement that will transfer control elsewhere or that will never return, not really for when you know that the list of enumerations (for example) is covered. It typically triggers undefined behavior. I would suggest to remove this line and just return error, or remove this line and log an error with `ereport`.

Reply:
I disagree. It is there for two reasons:

1. To safe-guard that the code above doesn't miss an enumeration or to catch that if upstream adds a new enumeration. Note that the compiler cannot really know since the above is a string-comparison in a for loop, not, e.g., a switch.
2. To let the people reading the code know that the one who wrote the code intentionally designed the code to not reach this point. Otherwise, someone would have asked to add a test case for this point based on the code coverage complaint.

If code coverage was smart enough, the macro would also help it to understand that this part of the code is not reachable and then it wouldn't complain.

I don't think we should spend time formulating errors that will never be thrown.

Reply:
Btw, in PostgreSQL, they use `pg_unreachable` exactly for the case you are arguing against: https://github.com/postgres/postgres/blob/e6ccd1ce1644d1b40b7981f8bc172394de524f99/src/backend/tcop/dest.c#L158